design/ceph/update-osds-in-parallel.md

```diff
@@ -0,0 +1,202 @@
+applications using the underlying Ceph storage. In some degraded cases, adding new storage may be
+necessary in order to allow currently-running Ceph OSDs to be updated without experiencing storage
+cluster downtime.
+
+This prioritization might delay updates significantly since adding OSDs not only adds capacity to 
+the Ceph cluster but also necessitates data rebalancing. Rebalancing generates data movement which
+needs to settle for updates to be able to proceed.
+
+
+## Technical implementation details
+
+STILL VERY MUCH WIP...
+
+### Definitions
+- **batch**: a list of OSDs that the operator wishes to update simultaneously
+- **query**: unit of work to determine if it is safe to update an OSD
+  - in practice, this is the Ceph command `ceph osd ok-to-stop <osd-id>...`
+- **update**: unit of work to update the OSD Deployment for an OSD
+- **batch query**: unit of work to determine if it is safe to update a batch of OSDs at once
+  - in practice, this is the Ceph command `ceph osd ok-to-stop <osd-id> <osd-id>...`
+- **batch update**: unit of work to update the OSD Deployments for a batch of OSDs at once
+
+### Workflow
+#### High level flow
+1. Build an "update queue" of OSDs in the cluster with failure domain information
+2. Based on the update strategy, find a batch of OSDs from the update queue that can be updated
+3. Remove OSDs in this batch from the update queue
+4. Update the batch of OSD Deployments
+5. If a Deployment fails to be updated, requeue it up to two times, or determine that it failed if
+   this was the third try
+6. If there are remaining OSDs in the update queue, repeat from step 2
+
+#### Build the update queue
+List all OSD Deployments belonging to the Rook cluster to use as the update queue. Go channels are
+likely not the best option since it may be necessary to select individual items from the queue at
+any position and not merely grab the top/bottom item.
+
+Also keep a record of the number of times each OSD has be queued to determine when Rook should stop
+attempting to update the OSD deployment. This can be attached to each OSD in the update queue
+via a struct, or it could be a map structure that can be easily looked up. Attaching via a struct
+would also allow easily attaching additional items (like a list of errors, for example) to OSDs in
+the future.
+
+`kubectl describe deployment rook-ceph-osd-0` on a test cluster yields ~8kB of text. If we assume
+that an OSD Deployment that is more complicated contains 10kB of data, listing all OSD Deployments
+for a cluster with 1000 OSDs would use approximately 10MB of RAM. An extremely large cluster with
+10000 OSDs would use approximately 100MB of RAM. This is likely acceptable for any modern system.
+There is an option in the future to build the update queue by listing OSDs in a paginated manner and
+building the queue out of the minimal set of information needed to perform the updates to reduce
+this usage.
+
+#### Get a batch of OSDs from the update queue
```

Comment:
To summarize the discussion in huddle and based on thoughts from Sage:
1. Base the decision to update purely on ok-to-stop
1. Using the CRUSH map, query ok-to-stop for a single OSD that is ready to be updated
1. If the single osd was ok, query for ok-to-stop for all OSDs that fall under the CRUSH hierarchy one level up. 
1. Continue moving up the CRUSH hierarchy until it is not ok-to-stop
1. Update all OSDs in parallel for the last level of the CRUSH hierarchy where it was ok
1. Return to step 2 until all OSDs are updated (or perhaps we timeout and requeue the entire reconcile)

Reply:
This makes sense. 

Questions:
1. will administrators want to specify a maxUpdatesInParallel option?
2. will administrators want to specify a sort of maximum hierarchy? (A user might want to enforce that we don't update at a higher level than rack, for example)
3. let's assume that there is one OSD on a host that can't be updated for some reason; should we try to update as many OSDs on the host as possible at the same time (i.e., all except the one) to do out best to avoid the case where we are doing updates one-OSD-at-a-time again?

> or perhaps we timeout and requeue the entire reconcile
Doesn't the mon health checker make sure that the mon cluster is healthy while we are doing OSD updates? It seems okay but strange to requeue the entire reconcile after an OSD update timeout. I would think it feels more natural to try to update all the OSDs in one go if at all possible.

Reply:
> This makes sense.
> 
> Questions:
> 
> 1. will administrators want to specify a maxUpdatesInParallel option?

Yes, I think the max in parallel option makes sense as mentioned in the other comment.

> 2. will administrators want to specify a sort of maximum hierarchy? (A user might want to enforce that we don't update at a higher level than rack, for example)

I don't think we need this limitation. If it's ok-to-stop a set of OSDs, Ceph is telling us it's ok. Not sure why the admin would care beyond this. Of course, we can always add such an option later if needed.

> 3. let's assume that there is one OSD on a host that can't be updated for some reason; should we try to update as many OSDs on the host as possible at the same time (i.e., all except the one) to do out best to avoid the case where we are doing updates one-OSD-at-a-time again?

This seems like an infrequent case. If a few OSDs need to be updated individually it will be fine. Other nodes would then be able to update all their OSDs, so overall the upgrade time should still be reasonable even if a few OSDs aren't ok-to-stop at a higher level.

> > or perhaps we timeout and requeue the entire reconcile
> 
> Doesn't the mon health checker make sure that the mon cluster is healthy while we are doing OSD updates? It seems okay but strange to requeue the entire reconcile after an OSD update timeout. I would think it feels more natural to try to update all the OSDs in one go if at all possible. And the other "child" CRDs have their own reconcile loops. Maybe this is a place where we can have some future plans to have a separate reconcile loop for OSDs...?

Actually I think we already have the behavior you're describing, and makes sense to keep it. If a single OSD cannot be updated, we continue to attempt all OSDs, but then requeue the reconcile at the end if any of the OSDs failed to update. 

A related behavior added recently is that at certain "checkpoints", the operator will cancel a reconcile if another event like update or delete was raised. This should be sufficient for cases to avoid the operator not responding to the other events.